libnmstate/nm/checkpoint.py

```diff
@@ -149,31 +117,127 @@ def _refresh_checkpoint_timeout(self, _user_data):
+        if self._dbuspath:
+            action = f"Destroy checkpoint {self._dbuspath}"
+            userdata = action
+            self._ctx.register_async(action)
+            try:
+                self._ctx.client.checkpoint_destroy(
+                    self._dbuspath,
+                    self._ctx.cancellable,
+                    self._checkpoint_destroy_callback,
+                    userdata,
+                )
+            except Exception as e:
+                raise NMCheckPointError(
+                    "Failed to destroy checkpoint {}: "
+                    "{}".format(self._dbuspath, e)
+                )
+            finally:
+                self._remove_checkpoint_refresh_timeout()
+            logging.debug(f"Checkpoint {self._dbuspath} destroyed")
+            self._ctx.wait_all_finish()
 
     def rollback(self):
-        try:
-            result = self._manager.interface.CheckpointRollback(self._dbuspath)
-        except dbus.exceptions.DBusException as e:
-            raise NMCheckPointError(str(e))
-        finally:
-            self._remove_checkpoint_refresh_timeout()
-        logging.debug(
-            "Checkpoint %s rollback executed: %s", self._dbuspath, result
-        )
-        return result
+        if self._dbuspath:
+            action = f"Rollback to checkpoint {self._dbuspath}"
+            userdata = action
+            self._ctx.register_async(action)
+            try:
+                self._ctx.client.checkpoint_rollback(
+                    self._dbuspath,
+                    self._ctx.cancellable,
+                    self._checkpoint_rollback_callback,
+                    userdata,
+                )
+            except Exception as e:
+                raise NMCheckPointError(
+                    "Failed to rollback checkpoint {}: {}".format(
+                        self._dbuspath, e
+                    )
+                )
+            finally:
+                self._remove_checkpoint_refresh_timeout()
+            logging.debug(f"Checkpoint {self._dbuspath} rollback executed")
+            self._ctx.wait_all_finish()
 
     @property
     def dbuspath(self):
         return self._dbuspath
+
+    def _checkpoint_create_callback(self, client, result, data):
+        try:
+            cp = client.checkpoint_create_finish(result)
+            self._dbuspath = cp.get_path()
+            if cp:
+                logging.debug(
+                    "Checkpoint {} created for all devices".format(
+                        self._dbuspath
+                    )
+                )
+                self._ctx.finish_async("Create checkpoint")
+            else:
+                error_msg = (
+                    f"dbuspath={self._dbuspath} "
+                    f"timeout={self._timeout} "
+                    f"callback result={cp}"
+                )
+                self._ctx.fail(
+                    NmstateLibnmError(f"Checkpoint create failed: {error_msg}")
```

Comment:
Should handle two cases:
 * NmstateConflictError for two+ checkpoint.
 * NmstatePermissionError for non-root.

Reply:
Done, this is now handing the permission and the conflict error on the try/except.